lib/caching.ts

```diff
@@ -0,0 +1,83 @@
+import { nrkRadio } from "./nrk/nrk.ts";
+import { parse } from "./parse.ts";
+import { Series, storage } from "./storage.ts";
+import * as datetime from "datetime";
+
+const SYNC_INTERVAL_HOURS = 1;
+
+async function initialFetch(options: { id: string }) {
+  const fromNrk = await nrkRadio.getSerieData(options.id);
+  const parsed = parse.series(fromNrk);
+
+  const stored = storage.write(parsed);
+  if (!stored) {
+    throw new Error(`Failed to store series ${options.id}`);
+  }
+
+  return parsed;
+}
+
+async function updateFetch(existingSeries: Series) {
+  const fromNrk = await nrkRadio.getSerieData(existingSeries.id);
+  const parsed = parse.series(fromNrk);
+
+  const newEpisodes = parsed.episodes.filter((episode) => {
+    return !existingSeries.episodes.find((serieEpisode) => serieEpisode.id === episode.id);
+  });
+
+  const updated = {
+    ...existingSeries,
+    lastFetch: new Date(),
+    /**
+     * Since we don't control the API,
+     * we should not make assumptions about the order,
+     * but rather sort the episode to what we want.
+     */
+    episodes: [...newEpisodes, ...existingSeries.episodes]
+      .sort((a, b) => a.date.getTime() > b.date.getTime() ? -1 : 1),
+  };
+
+  const updateSuccessful = await storage.write(updated);
+  if (!updateSuccessful) {
+    throw new Error(`Failed to update series ${existingSeries.id}`);
+  }
+
+  return updated;
+}
+
+async function getSeries(options: { id: string }): Promise<Series> {
+  const seriesFromStorage = await storage.read(options);
+
+  /**
+   * We don't have the feed in storage,
+   * and we need to fetch it for the first time.
+   */
+  if (seriesFromStorage === null) {
+    // TODO: schedule a job to fetch the entire backlog https://github.com/olaven/NRSS/issues/24
+    return await initialFetch(options);
+  }
+
+  const timeSinceLastFetch =
+    datetime.difference(seriesFromStorage.lastFetchedAt, new Date(), { units: ["hours"] }).hours;
+
+  // we have the feed in storage and it's not too old
+  if (
+    seriesFromStorage !== null &&
+    timeSinceLastFetch !== null &&
+    timeSinceLastFetch !== undefined &&
+    timeSinceLastFetch <= SYNC_INTERVAL_HOURS
+  ) {
+    return seriesFromStorage;
+  }
```

Comment:
Could this be simplified, or is the verbosity intentional?

```suggestion
  if ( seriesFromStorage && timeSinceLastFetch && timeSinceLastFetch <= SYNC_INTERVAL_HOURS) {
    return seriesFromStorage;
  }
```

Reply:
The verbosity is intentional, but I am aware that it's slightly unconventional. When JavaScript is being strange, I prefer to be explicit about it.

Reply:
I'm resolving this for now, but I'm happy to keep the conversation going if you want.

Reply:
Sounds good! 👌

Reply:
It also occurred to me that in this case, using the proposed varient would lead to a (very well hidden) bug. Whenever timeSinceLastFetch would be `0`, the check would fail (even though it was defined) because `0` is falsy. Since the limit is `1`, an update would always be triggered. Makes sense?

Reply:
(and I believe it would be `0` for all moments up until 1 hour has been reached)